format/spec.md

```diff
@@ -1297,54 +1308,56 @@ Example
+| **`fixed(L)`**       | **`JSON string`**                                            | `"000102ff"`                               | Stored as a hexadecimal string                                                                                                                                                                                                                                                                           |
+| **`binary`**         | **`JSON string`**                                            | `"000102ff"`                               | Stored as a hexadecimal string                                                                                                                                                                                                                                                                           |
+| **`struct`**         | **`JSON object by field ID`**                                | `{"1": 1, "2": "bar"}`                     | Stores struct fields using the field ID as the JSON field name; field values are stored using this JSON single-value format                                                                                                                                                                              |
+| **`list`**           | **`JSON array of values`**                                   | `[1, 2, 3]`                                | Stores a JSON array of values that are serialized using this JSON single-value format                                                                                                                                                                                                                    |
+| **`map`**            | **`JSON object of key and value arrays`**                    | `{ "keys": ["a", "b"], "values": [1, 2] }` | Stores arrays of keys and values; individual keys and values are serialized using this JSON single-value format                                                                                                                                                                                          |
+| **`variant`**        | **`Same JSON representation in this table for stored type`** | `null`, `true`, `{"1": 1, "2": "bar"}`      | The JSON representation matches the format shown in this table for the type stored in the Variant.                                                                                                                                                                                                       |
```

Comment:
I think this is insufficient because it loses type information. For example, if there is a Variant that contains a timestamp, the timestamp's type is lost. This also appears to use the struct representation for an object, but a Variant has no field IDs.

I think that we probably do want to use a JSON representation, but we will need to make sure that it aligns with the JSON conversion defined in the Variant spec (https://github.com/apache/parquet-format/pull/461/files#diff-80a56bf0d841087ab5038020e4b78119d6ceb44684719ba4d3a6e22effb36eb9R459) and that we have defined requirements for recovering the types.

Note that the JSON conversion in the Variant spec differs from this section:
* Binary values are hexadecimal strings here (:facepalm:) and base64 in Variant
* Decimal values are strings here and numbers in Variant

We may want to not allow Variant in JSON because of the type loss. Or we may want to specify a subset that can be recovered, like booleans, integers, floats, strings, and arrays. There are two places where this section is used: in the REST protocol for sending lower/upper bounds and in default values. We don't really need default values for variants (but would have to disallow them) and lower/upper bounds would generally work with a subset.

The other option is to encode `variant` as a base64 binary value. That is a bit ugly in JSON, but it is not lossy.

Reply:
I don't exactly understand how this gets used. I will make changes to say "The JSON representation with the encoding defined in the https://github.com/apache/parquet-format/pull/461/files#diff-80a56bf0d841087ab5038020e4b78119d6ceb44684719ba4d3a6e22effb36eb9R459 of Variant spec". Would that work for lower/upper bounds usage? 

I notice that SingleValueParser.fromJson(Type type, JsonNode defaultValue) and toJson(Type type, Object defaultValue, JsonGenerator generator) would error out for Variant. Is that sufficient to disallow defaultValue?